api/grpcserver/post_service.go

```diff
@@ -0,0 +1,135 @@
+package grpcserver
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	pb "github.com/spacemeshos/api/release/go/spacemesh/v1"
+	"go.uber.org/zap"
+)
+
+// PostService is a grpc server that PoST nodes can connect to in order to register.
+// The bidirectional stream established between the node and the PoST node can be used
+// to send challenges and receive proofs.
+type PostService struct {
+	log       *zap.Logger
+	callbacks []postConnectionListener
+
+	// connections is a map of IDs -> channel of commands for every connected node.
+	// the node calls functions on this service which build API requests for every
+	// connected node and send them on the channel. The PoST node receives the requests
+	// and sends responses back on the response channel of the command.
+	conMtx sync.Mutex
+	con    chan postCommand
+	client *postClient
+}
+
+// RegisterService registers this service with a grpc server instance.
+func (s *PostService) RegisterService(server *Server) {
+	pb.RegisterPostServiceServer(server.GrpcServer, s)
+}
+
+// NewPostService creates a new grpc service using config data.
+func NewPostService(log *zap.Logger, callbacks ...postConnectionListener) *PostService {
+	return &PostService{
+		log:       log,
+		callbacks: callbacks,
+	}
+}
+
+// Register is called by the PoST service to connect with the node.
+// It creates a bidirectional stream that is kept open until either side closes it.
+// The other functions on this service are called by services of the node to send
+// requests to the PoST node and receive responses.
+func (s *PostService) Register(stream pb.PostService_RegisterServer) error {
+	ch := make(chan postCommand)
+	s.setConnection(ch)
+	defer s.dropConnection()
+
+	for {
+		select {
+		case <-stream.Context().Done():
+			return stream.Context().Err()
```

Comment:
Both the sending and the closing of the `ch` channel are protected by the same mutex:

```go
func (s *PostService) dropConnection() error {
	s.conMtx.Lock()
	defer s.conMtx.Unlock()

	if s.con == nil {
		return fmt.Errorf("connection not established")
	}

	close(s.con)
	s.con = nil
	s.client.Close()

	for _, cb := range s.callbacks {
		cb.Disconnected(s.client)
	}

	s.client = nil
	return nil
}
```

```go
func (s *PostService) Send(ctx context.Context, cmd postCommand) error {
	s.conMtx.Lock()
	defer s.conMtx.Unlock()

	if s.con == nil {
		return fmt.Errorf("connection not established")
	}

	select {
	case <-ctx.Done():
		return ctx.Err()
	case s.con <- cmd:
		return nil
	}
}
```

I don't think a deadlock as you describe is possible in this code. The reason `Send` is mutually exclusive is to avoid two parts of the node sending a command to the post service before the response is received.

Reply:
I removed `PostService::Send` in favour of passing `s.con` directly to the `PostClient`. Every time it tries to send on this channel it will also check `pc.closed` in case it has been closed already (an no one is receiving on `s.con` anymore). This should avoid deadlocks.